onnx/common/status.h

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Facebook Inc. and Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+#include <memory>
+#include <string>
+
+namespace ONNX_NAMESPACE {
+namespace Common {
+
+enum StatusCategory {
+  NONE = 0,
+  CHECKER = 1,
+  OPTIMIZER = 2,
+};
+
+enum StatusCode {
+  OK = 0,
+  FAIL = 1,
+  INVALID_ARGUMENT = 2,
+  INVALID_PROTOBUF = 3,
+};
+
+class Status {
+ public:
+  Status() noexcept {}
+
+  Status(StatusCategory category, int code, const std::string& msg);
+
+  Status(StatusCategory category, int code);
+
+  Status(const Status& other) {
+    *this = other;
+  }
+
+  void operator=(const Status& other) {
+    if (&other != this) {
+      if (nullptr == other.state_) {
+        state_.reset();
+      } else if (state_ != other.state_) {
+        state_.reset(new State(*other.state_));
+      }
+    }
+  }
+
+  Status(Status&& other) = default;
+  Status& operator=(Status&& other) = default;
+  ~Status() = default;
+
+  bool IsOK() const noexcept;
+
+  int Code() const noexcept;
+
+  StatusCategory Category() const noexcept;
+
+  const std::string& ErrorMessage() const;
+
+  std::string ToString() const;
+
+  bool operator==(const Status& other) const {
+    return (this->state_ == other.state_) || (ToString() == other.ToString());
```

Comment:
Better to implement `==` for class `State` so we don't need to construct the string representation of status.

Reply:
will check this comment later.